recon/src/recon.rs

```diff
@@ -605,355 +665,37 @@ where
-            .await?;
-        let head: Vec<K> = range.by_ref().take(SPLIT_THRESHOLD).collect();
-
-        if head.len() < SPLIT_THRESHOLD {
-            trace!("sending all keys");
-            for key in head {
-                self.keys.push(key.to_owned());
-                self.hashes.push(H::identity());
-            }
-        } else {
-            let mid_key = local_store.middle(left_fencepost, right_fencepost).await?;
-            trace!(?mid_key, "splitting on key");
-            if let Some(mid_key) = mid_key {
-                self.keys.push(mid_key.to_owned());
-                self.hashes
-                    .push(local_store.hash_range(left_fencepost, &mid_key).await?.hash);
-
-                self.keys.push(right_fencepost.to_owned());
-                self.hashes.push(
-                    local_store
-                        .hash_range(&mid_key, right_fencepost)
-                        .await?
-                        .hash,
-                );
-            } else {
-                bail!("unable to find a split key")
-            };
-        }
-        Ok(())
-    }
-
-    fn bound(&self, range: &RangeOpen<K>) -> BoundedMessage<K, H> {
-        // TODO: Can we do this without allocating?
-        // Some challenges that make this hard currently:
-        //  1. The process_messages method iterates over the keys twice
-        //  2. We have to keep keys and hashes properly aligned
-
-        let mut hashes = self.hashes.iter();
-        let keys: Vec<K> = self
-            .keys
-            .iter()
-            .skip_while(|key| {
-                if *key <= &range.start {
-                    // Advance hashes
-                    hashes.next();
-                    true
-                } else {
-                    false
-                }
-            })
-            .take_while(|key| *key < &range.end)
-            .cloned()
-            // Collect the keys to ensure side effects of iteration have been applied.
-            .collect();
-
-        // Collect the hashes up to the last key.
-        let hashes = if keys.len() <= 1 {
-            Vec::new()
-        } else {
-            hashes.take(keys.len() - 1).cloned().collect()
-        };
-
-        BoundedMessage { keys, hashes }
-    }
-
-    fn range(&self) -> RangeOpen<K> {
-        RangeOpen {
-            start: self.start.to_owned().unwrap_or_else(|| K::min_value()),
-            end: self.end.to_owned().unwrap_or_else(|| K::max_value()),
-        }
-    }
+/// Represents a synchronization unit, a pair of keys and the hash of values between the keys
+/// (exclusive of the keys).
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct Range<K, H> {
+    /// First key in the range
+    /// This key may be a fencepost, meaning its not an actual key but simply a boundary.
+    pub first: K,
+    /// Hash of all keys in between first and last exclusive.
+    /// Hash is zero if there are no keys within the range.
+    pub hash: HashCount<H>,
```

Comment:
Should this be an Optional HashCount

core::RangeOpen and recon:Range feel like they are similar do we want difrent types for ranges based on wether the hash and count are known?

Obviously we need the hash if we want to send it across the wire.
Maybe hash_range should take a RangeOpen and return a Range?

Reply:
I actually started with an Option<HashCount> but it turns out since Hash has a well defined zero value that always expecting the hash was better. This way sqlite computing a sum doesn't have to have logic to check for a zero and return a None instead. In other words there is no meaningful difference between a None hash and a zero hash so an Option only adds unneeded complexity.

Having two types works well for me. However we could maybe have better names? And yes a nice refactor later would be to change all the method that accept left_fencepost and right_fencepost to simply accept a RangeOpen.

Reply:
The None would be unknown not zero. But look at it the second time having one type for an open range and different one for a range with its hash and count is clearer. There is no one place where we have ranges but don't know if we retrieved the hash and count yet.